In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os


In [2]:
#Creating a path for this project
path=r'C:\\Users\\Admin\\05-01-2023 Instacart Basket Analysis'
print(path)

C:\\Users\\Admin\\05-01-2023 Instacart Basket Analysis


In [3]:
df_orders_products_merged= pd.read_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merge.pkl'))

In [21]:
df_orders_products_merged.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices exists    price_range_loc  
0         Soda        77              7     9.0   both  Mid-range product  
1         Soda        77              7     9.0   both  Mid-range product  
2         Soda        77              7     9.0   both  Mid-range product  
3         Soda        77              7     9.0   both  Mid-range product  
4         Soda        77              7     9.0   both  Mid-range product

In [5]:
df = df_orders_products_merged[:1000000]

1. If-Statement using function

In [6]:
# Defining a function

def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'


In [7]:
df['price_range'] = df.apply(price_label, axis=1)

C:\Users\Admin\AppData\Local\Temp\ipykernel_11532\2848743372.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [8]:
df['price_range'].value_counts(dropna = False)

Mid-range product    709284
Low-range product    280808
High range             9908
Name: price_range, dtype: int64

In [9]:
df['prices'].max()

14900.0

Aplying loc() function to dataset

In [10]:
df_orders_products_merged.loc[df_orders_products_merged['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [11]:
df_orders_products_merged.loc[(df_orders_products_merged['prices'] <= 15) & (df_orders_products_merged['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 


In [12]:
df_orders_products_merged.loc[df_orders_products_merged['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [13]:
df_orders_products_merged['price_range_loc'].value_counts(dropna = False)

Mid-range product     2153643
Low-range product      995802
High-range product      40269
Name: price_range_loc, dtype: int64

3. If-Statements with For-Loops

#Rename'order_dow' to 'orders_day_of_the_week'\n",


In [25]:
df_orders_products_merged = df_orders_products_merged.rename(columns={'order_dow': 'orders_day_of_week'})
df_orders_products_merged.head()


order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered _merge product_name  aisle_id  department_id  prices exists  \
0          0   both         Soda        77              7     9.0   both   
1          1   both         Soda        77              7     9.0   both   
2          1   both         Soda        77              7     9.0   both   
3          1   both         Soda        77              7     9.0   both   
4          1   both         Soda        77              7     9.0   both   

     price_range_loc  
0  Mid-range product  
1  Mid-range product  
2  Mid-range product  
3  Mid-range product  
4  Mid-range product

In [26]:
df_orders_products_merged['orders_day_of_week'].value_counts(dropna = False)

0    605483
1    548057
6    446607
5    420357
2    414852
3    379676
4    374682
Name: orders_day_of_week, dtype: int64

In [34]:
# Creating the vector where the outputs from the loop will be stored
result2 = []

In [32]:
df_orders_products_merged['busiest_days'] = result2

In [33]:
# Checking value counts
df_orders_products_merged['busiest_days'].value_counts(dropna = False)

Regularly days    1281816
Busiest days      1153540
Slowest days       754358
Name: busiest_days, dtype: int64

In [30]:
# Creating the loop
for value in df_orders_products_merged['orders_day_of_week']:
  if value == 0 or value ==1:
    result2.append('Busiest days')
  elif value == 4 or value == 3:
    result2.append('Slowest days')
  else:
    result2.append('Regularly days')

In [35]:
df_orders_products_merged.shape

(3189714, 18)

Q2. Suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). At the same time, they’d also like to know the two slowest days. Create a new column for this using a suitable method.

In [36]:
df_orders_products_merged.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered _merge product_name  aisle_id  department_id  prices exists  \
0          0   both         Soda        77              7     9.0   both   
1          1   both         Soda        77              7     9.0   both   
2          1   both         Soda        77              7     9.0   both   
3          1   both         Soda        77              7     9.0   both   
4          1   both         Soda        77              7     9.0   both   

     price_range_loc    busiest_days  
0  Mid-range product  Regularly days  
1  Mid-range product    Slowest days  
2  Mid-range product    Slowest days  
3  Mid-range product    Slowest days  
4  Mid-range product    Slowest days

#Q3. Check the values of this new column for accuracy. Note any observations in markdown format.

In [38]:
counts = df_orders_products_merged['order_hour_of_day'].value_counts().rename_axis('hour').reset_index(name='counts')
counts

hour  counts
0     10  270097
1     11  269078
2     14  262566
3     13  261539
4     15  261508
5     12  260701
6     16  248333
7      9  240820
8     17  209187
9      8  168824
10    18  162459
11    19  125314
12    20   96570
13     7   89322
14    21   79597
15    22   62771
16    23   38407
17     6   28002
18     0   19248
19     1   10600
20     5    8609
21     2    6128
22     4    5534
23     3    4500

In [39]:
orders = list(df_orders_products_merged['order_hour_of_day'].value_counts().quantile([0.25, 0.5, 0.75]))
orders

[25813.5, 110942.0, 251425.0]

In [17]:
# Checking value counts
df_orders_products_merged['busiest_days'].value_counts(dropna = False)

Regularly days    618611
Busiest days      549404
Slowest days      365052
Name: busiest_days, dtype: int64

Step 3: The total number of busiest days the sum of 0 & 1 Regualrly days(618611) and slowest day (365052)."


Grouping into Most orders, Average orders and Fewest orders, according to percentiles


Lists: I created thre lists for the three different labels: “Most orders,” “Average orders,” and “Fewest orders"

In [40]:
# Fewest orders list
few = list(counts.loc[counts['counts'] <= orders[0]]['hour'])

In [44]:
# Average orders list
avg_ords = list(counts[(counts['counts'] > orders[0]) & (counts['counts'] < orders[2])]['hour'])

In [45]:
# Most orders list
Mos_ords = list(counts.loc[counts['counts'] >= orders[2]]['hour'])

Index: I created three indexs for the three different labels: “Most orders,” “Average orders,” and “Fewest orders"

In [46]:
index_few = df_orders_products_merged['order_hour_of_day'].isin(few)

In [47]:
index_avg_ords = df_orders_products_merged['order_hour_of_day'].isin(avg_ords)


In [48]:
index_Mos_ords = df_orders_products_merged['order_hour_of_day'].isin(Mos_ords)

Create a new column containing these labels called “busiest_period_of_day.”

In [49]:
df_orders_products_merged.loc[index_few,'busiest_period_of_day'] = 'Fewest orders'

In [50]:
df_orders_products_merged.loc[index_avg_ords,'busiest_period_of_day'] = 'Average orders'

In [51]:
df_orders_products_merged.loc[index_Mos_ords,'busiest_period_of_day'] = 'Most orders'

In [52]:
df_orders_products_merged.head()


order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered _merge product_name  aisle_id  department_id  prices exists  \
0          0   both         Soda        77              7     9.0   both   
1          1   both         Soda        77              7     9.0   both   
2          1   both         Soda        77              7     9.0   both   
3          1   both         Soda        77              7     9.0   both   
4          1   both         Soda        77              7     9.0   both   

     price_range_loc    busiest_days busiest_period_of_day  
0  Mid-range product  Regularly days        Average orders  
1  Mid-range product    Slowest days        Average orders  
2  Mid-range product    Slowest days           Most orders  
3  Mid-range product    Slowest days        Average orders  
4  Mid-range product    Slowest days           Most orders

In [53]:
df_orders_products_merged.shape

(3189714, 19)

Q5. Print the frequency for this new column.

In [54]:
df_orders_products_merged['busiest_period_of_day'].value_counts()

Most orders       1585489
Average orders    1549606
Fewest orders       54619
Name: busiest_period_of_day, dtype: int64

Q7. Export your dataframe as a pickle file (since you added new columns) and store it correctly in your “Prepared Data” folder.

In [55]:
# Export to PKL
df_orders_products_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_update.pkl'))